### DISCUSSION TITLE
Android app

### DISCUSSION THREAD

**Discussion by WimDeMeester** [WimDeMeester/eFinder#54] [Ideas]
Discussion started by @AstroKeith , in PR #47 (See https://github.com/WimDeMeester/eFinder/pull/47#issuecomment-1585743139)

Couldnt resist taking a look at the Android App challenge.

Turns out it’s very similar to Java, with the complications mostly in the UI. I’m running Android Studio and its pretty impressive - including a pretty good device emulator. After a day of learning I’m getting my own basic Apps running.

The big challenge will be the communication with the Pi. Lots of options and none simple. So some questions for you…

What would be your preferred connection method?

USB
Bluetooth (slow)
Wifi peer to peer
Wifi Access Point (ie Pi joins the Device, or vice versa
Wifi infrastructure mode - use home router, portable router or Nexus DSC as a router.

Your choice may well be determined for you by your other connections. I’m using USB from Pi to Nexus DSC, Nexus DSC wifi to SkySafari, so I have an unused Pi wifi. I can also use the Nexus DSC as a wifi router but likely to get problems running SkySafari in parallel with my new App down the same device wifi connection. The Nexus DSC Pro cant act as a router.

Rather than try and reproduce the VNCGUI screen on the new App, I’m thinking of making an optional ‘plug-in’ to the hand box eFinder version. It would be just a basic display of the captured image (useful for focusing and parameter tweaking), and a display of an annotated image. Thus the ‘buttons’ on the app would be just two - for load last image or capture new image, and a series of check boxes to decide the annotation layers (including none).

Any though?

**Comment by WimDeMeester** [comment]
Hi @AstroKeith,

Very nice that you have enough time to check the possibility for an Android app. 
I don't think that a USB connection is a good option.  This means that the Phone or Tablet needs to be connected all the time to the Raspberry PI.  I like to have my tablet on a seperate table and not attached to my telescope. 
I guess the WiFi option is the best one.  My tablet is anyway connected to the tablet for using SkySafari.  Is it possible to connect the tablet directly to the Raspberry PI as well?

**Comment by AstroKeith** [comment]
Having spent a weekend indoors hiding from the heatwave, I can see that establishing comms between an Android device and the Pi is going to be a challenge and a half! Exchanging messages is foreseeable but transmitting whole images might take some real learning!

If you have your tablet connected to the Nexus DSC in Access Point mode, then it depends which Nexus DSC you have! The Original Nexus DSC can act as a wifi router, which means if we configure the Pi to connect to the Nexus DSC, then the App would be able to talk to the Pi.

If its a 'Pro' then the Nexus has a bug and cannot act as a router.

A way around this is to have a travel wifi router (assuming you are away from home), and get all devices (Pi, Nexus DSC Pro & tablet) to connect to the travel router.

My star party sessions include a lot of debugging and experimentation, so I am now in the habit of taking a wifi router with me which allows me to connect 'everything' and sit inside my caravan and 'play'. The wifi router also has a 4G data SIM card slot.

**Reply by WimDeMeester** [reply]
Mine is indeed a Nexus Pro...  I guess I will need to add a travel router to my list of things to buy.  Do you have a suggestion?

**Reply by AstroKeith** [reply]
Because I have mains power in my caravan I use the model https://www.amazon.co.uk/TL-MR6400-Unlocked-Configuration-Required-External/dp/B016ZWXYXG/ref=sr_1_3?crid=HDNNJSDOXN2F&keywords=tp+link+sim+router&qid=1686575428&sprefix=tp+link+sim+router%2Caps%2C89&sr=8-3

But these are good I believe .
https://www.amazon.co.uk/D-Link-Mbps-Wi-Fi-Hotspot-DWR-932/dp/B00OVJDCQ0/ref=sr_1_3?crid=3VVSHLQ0QS2MF&keywords=portable+router&qid=1686574774&sprefix=portable+router%2Caps%2C84&sr=8-3

https://www.amazon.co.uk/Tenda-WLAN-Router-4G185-v2-0/dp/B08FC72QD8/ref=sr_1_10?crid=3VVSHLQ0QS2MF&keywords=portable%2Brouter&qid=1686574774&sprefix=portable%2Brouter%2Caps%2C84&sr=8-10&th=1

**Reply by WimDeMeester** [reply]
I suppose these routers also work without a SIM card (so no 4G)?

**Reply by AstroKeith** [reply]
My mains powered one does!

**Comment by WimDeMeester** [comment]
> Exchanging messages is foreseeable but transmitting whole images might take some real learning!

I think that this can be solved in a not so difficult way, as you have access to the Raspberry Pi and the Android App.  Personally, I would set up a webserver on the Raspberry PI (I would just serve the directory where the images are saved).  As the Android App is on the same network, you can just download the images from the website.

**Reply by WimDeMeester** [reply]
@AstroKeith : If you also think that this is a viable option, I can try to find out how to set up the web server and send you the instructions.

**Reply by AstroKeith** [reply]
Yes it probably is the best way to go. I have a web server on my ScopeDog (allows easy set up of parameters.) I'll dig out how I did that one! I might revert to your offer though!
